server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -90,15 +92,25 @@ abstract class AbstractSearchAsyncAction<Result extends SearchPhaseResult> exten
     private final Object shardFailuresMutex = new Object();
     private final AtomicBoolean hasShardResponse = new AtomicBoolean(false);
     private final AtomicInteger successfulOps = new AtomicInteger();
-    private final AtomicInteger skippedOps = new AtomicInteger();
     private final SearchTimeProvider timeProvider;
     private final SearchResponse.Clusters clusters;
 
     protected final GroupShardsIterator<SearchShardIterator> toSkipShardsIts;
     protected final GroupShardsIterator<SearchShardIterator> shardsIts;
     private final SearchShardIterator[] shardIterators;
-    private final int expectedTotalOps;
-    private final AtomicInteger totalOps = new AtomicInteger();
+
+    private static final VarHandle OUTSTANDING_SHARDS;
+
+    static {
+        try {
+            OUTSTANDING_SHARDS = MethodHandles.lookup().findVarHandle(AbstractSearchAsyncAction.class, "outstandingShards", int.class);
```

Comment:
The constructor here is confusingly expensive in profiling, maybe starting to relax the ordering a little will help so I figured I'd start here while I'm changing this anyway and at least go to just release style semantics in the constructor.

Reply:
I understand that skipped shards do not need their own counter as they are known in advance. I also agree that moving to counting each shard as 1 is simpler, and then there is no need to compare expected ops with total ops. 

Can we do without the `VarHandle` and use a simple `AtomicInteger` that we decrement each time though? What makes the var handle needed in this scenario?

Reply:
Right, the var handle came out of a version of this change that went a little further lets say :) with the way it works now, the atomic integer might also be faster, no need to put the changing counter in a cacheline with other read-only fields :) => atomic int it is